File: CS/PS/복습/progr_lv1/키패드누르기.py
# 16:56 ~ 17:10
def solution(numbers, hand):
    answer = ''
    # 4 by 3
    pad = {
        1 : [0,0],
        2 : [0,1],
        3 : [0,2],
        4 : [1,0],
        5 : [1,1],
        6 : [1,2],
        7 : [2,0],
        8 : [2,1],
        9 : [2,2],
        '*' : [3,0],
        0 : [3,1],
        '#' : [3,2]
        
    }
    left_pos = pad['*']
    right_pos = pad['#']
    for num in numbers:
        if (num == 1 or num == 4 or num == 7):
            answer += "L"
            left_pos = pad[num]
        elif (num  == 3 or num == 6 or num == 9):
            answer += "R"
            right_pos = pad[num]
        elif (num  == 2 or num == 5 or num == 8 or num == 0):
            # 이동거리 = > | key_pos_x - hand_pos_x | + |y거리| -> 같을 때 손잡이방향 우선
            mov_left_x = abs(pad[num][0] - left_pos[0])
            mov_left_y = abs(pad[num][1] - left_pos[1])
            mov_right_x = abs(pad[num][0] - right_pos[0])
            mov_right_y = abs(pad[num][1] - right_pos[1])
            
            if (mov_left_x + mov_left_y > mov_right_x + mov_right_y) : # left가 더 멀때
                answer += "R"
                right_pos = pad[num]
            elif mov_left_x + mov_left_y < mov_right_x + mov_right_y:
                answer += "L"
                left_pos = pad[num]
            else :
                if hand == "right":
                    answer += "R"
                    right_pos = pad[num]
                else :
                    answer += "L"
                    left_pos = pad[num]
            
            
        
    return answer

File: CS/PS/Site/programmers/lv1/키패드누르기.js
function findPos(numbers,keyPad){
  for(let [row, arr] of keyPad.entries()){
    let colPos = arr.indexOf(numbers);
    if(colPos > -1){
      return [row, colPos];
    }
  }
}

function solution(numbers, hand) {
  var answer = '';
  let keyPad = [
    [1,2,3],
    [4,5,6],
    [7,8,9],
    ['*',0,'#']
  ];
  
  let leftCurr = [3,0];
  let rightCurr = [3,2];
  let length = numbers.length;
  for(let i = 0; i < length; i++){
    let currNum = numbers[i];
    let numPos = findPos(numbers[i], keyPad);
    if(numbers[i] === 1 || numbers[i] === 4 || numbers[i] === 7){
      leftCurr = numPos;
      answer += 'L';
      continue;
    } else if (numbers[i] === 3 || numbers[i] === 6 || numbers[i] === 9){
      rightCurr = numPos;
      answer += 'R';
      continue;
    } else {
      //2580일때
      let leftGap = Math.abs(leftCurr[0] - numPos[0]) + Math.abs(leftCurr[1] - numPos[1]);
      let rightGap = Math.abs(rightCurr[0] - numPos[0]) + Math.abs(rightCurr[1] - numPos[1]);
      // console.log('numbers[i]: ', numbers[i], 'left Gap : ', leftGap,  'prevLeftPos', keyPad[leftCurr[0]][leftCurr[1]], 'numPos: ',numPos, 'leftCurr: ',leftCurr);
      // console.log( 'numbers[i]: ',numbers[i],'right Gap: ' ,rightGap,  'prevRighPos', keyPad[rightCurr[0]][rightCurr[1]], 'numPos: ',numPos, 'rightCurr: ',rightCurr);
      if (leftGap > rightGap){
        //left가 더 멀 때
        rightCurr = numPos;
        answer += 'R';
      }
      else if (leftGap < rightGap){
        leftCurr = numPos;
        answer += 'L';
      } else{
        if(hand === "right"){
          rightCurr = numPos;
          answer += 'R';
        } else {
          leftCurr = numPos;
          answer += 'L';
        }
      }

    }
  }
  return answer;
}

// console.log(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"));
console.log(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2],"left"));
// console.log(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0],"right"));

